rust/datafusion/src/execution/physical_plan/projection.rs

```diff
@@ -0,0 +1,238 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines the projection execution plan. A projection determines which columns or expressions
```

Comment:
Hmm. Not sure if I understand this. Shouldn't this be part of logical query optimization?

Reply:
Projection can happen in many places in a query plan e.g. `SELECT sqrt(x) FROM (SELECT MAX(a) AS x FROM foo)`. This is separate from the projection that gets pushed down to the data source.

Reply:
I see. I thought this can be done by a separate analysis on the logical plan though (in Hive this is how it is done with column pruning).